src/subhalo_merge.cpp

```diff
@@ -9,312 +9,212 @@
-  for (HBTInt i = 0; i < Subhalos.size(); i++)
-  {
-    Helpers[i].BuildPosition(Subhalos[i]);
-    Helpers[i].BuildVelocity(Subhalos[i]);
-  }
-}
-void FillHelpers(vector<SubHelper_t> &Helpers, const vector<Subhalo_t> &Subhalos)
-{
-  FillHostTrackIds(Helpers, Subhalos);
-  FillCores(Helpers, Subhalos);
-}
-void SubhaloSnapshot_t::MergeSubhalos()
-{
-  HBTInt NumHalos = MemberTable.SubGroups.size();
-  vector<SubHelper_t> Helpers(Subhalos.size());
-  int isnap = GetSnapshotIndex();
-
-#pragma omp parallel
-  {
-    GlueHeadNests();
-    FillHelpers(Helpers, Subhalos);
+    if (HBTConfig.PeriodicBoundaryOn)
+      CoreComovingPosition[dim] += origin[dim];
 
-    DetectTraps(Subhalos, Helpers, isnap);
-  }
+    pos2[dim] -= pos[dim] * pos[dim];
 
-  if (HBTConfig.MergeTrappedSubhalos)
-  {
-    /* Remove particles from merged tracks, and store when this happened */
-#pragma omp parallel for schedule(dynamic, 1)
-    for (HBTInt grpid = 0; grpid < NumHalos; grpid++)
-      if (MemberTable.SubGroups[grpid].size())
-        MergeRecursive(MemberTable.SubGroups[grpid][0]);
-
-        /* Subgroups receiving particles from merged ones are subject to unbinding*/
-#pragma omp parallel for schedule(dynamic, 1) if (ParallelizeHaloes)
-    for (HBTInt subid = 0; subid < Subhalos.size(); subid++)
-      if (Helpers[subid].IsMerged)
-        Subhalos[subid].Unbind(*this);
-
-        /* Truncate the source of the subhaloes we just updated. */
-#pragma omp parallel for
-    for (HBTInt subid = 0; subid < Subhalos.size(); subid++)
-      if (Helpers[subid].IsMerged)
-        Subhalos[subid].TruncateSource();
+    vel[dim] /= msum;
+    vel2[dim] /= msum;
+    CorePhysicalVelocity[dim] = vel[dim];
+    vel2[dim] -= vel[dim] * vel[dim];
   }
 
-  Helpers.clear();
-#pragma omp parallel
-  UnglueHeadNests();
-}
-
-void SubhaloSnapshot_t::MergeRecursive(HBTInt subid)
-{
-  auto &sat = Subhalos[subid];
-  for (auto &&nestid : sat.NestedSubhalos)
-    MergeRecursive(nestid);
-  if (sat.IsTrapped() && sat.IsAlive())
-  {
-    sat.MergeTo(Subhalos[sat.SinkTrackId]);
-    sat.SnapshotIndexOfDeath = GetSnapshotIndex();
-  }
+  CoreComovingSigmaR = sqrt(pos2[0] + pos2[1] + pos2[2]);
+  CorePhysicalSigmaV = sqrt(vel2[0] + vel2[1] + vel2[2]);
 }
 
-/*
-struct ParticleHasher_t //to be passed as a template parameter to unordered_set
-{
-  size_t operator()(const Particle_t& p) const
-  {
-    return p.Id;
-  }
-};
-*/
 void Subhalo_t::MergeTo(Subhalo_t &host)
 {
   if (Nbound <= 1)
     return; // skip orphans and nulls
 
 #ifndef INCLUSIVE_MASS
-  HBTInt np_max = host.Particles.size() + Particles.size();
+  /* We pass the particles from this subhalo to the one it merged with. We only
+   * do the bound subset, since the unbound subset has already made its way onto
+   * the upper hierarchy when this subhalo was subject to unbinding.  */
+  HBTInt np_max = host.Particles.size() + Nbound;
+
+  /* TODO: check whether this exclusivity list is required. If everything is
```

Comment:
Do you want to keep this? I agree it shouldn't be needed

Reply:
Moved to a DEBUG #ifdef block